compiler-core/src/analyse/name.rs

```diff
@@ -0,0 +1,116 @@
+use ecow::EcoString;
+
+use crate::ast::{ArgNames, SrcSpan};
+
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+type BadName = (SrcSpan, EcoString);
+
+pub fn check_valid_name(
+    location: SrcSpan,
+    name: &EcoString,
+    kind: BadNameKind,
+) -> Option<(Error, BadName)> {
+    check_bad_name(location, name, kind)
+        .map(|error| (error, (location, EcoString::from(name.to_snake_case()))))
+}
+
+pub fn check_valid_discard_name(location: SrcSpan, name: &EcoString) -> Option<(Error, BadName)> {
+    check_bad_name(location, name, BadNameKind::Discard).map(|error| {
+        (
+            error,
+            (
+                location,
+                EcoString::from(format!("_{}", name.to_snake_case())),
+            ),
+        )
+    })
+}
+
+pub fn check_valid_upname(
+    location: SrcSpan,
+    name: &EcoString,
+    kind: BadNameKind,
+) -> Option<(Error, BadName)> {
+    if name.contains('_') {
+        return Some((
+            Error::BadName {
+                location,
+                kind,
+                name: name.clone(),
+            },
+            (location, EcoString::from(name.to_upper_camel_case())),
+        ));
+    }
+    None
+}
+
+pub fn check_valid_argument(names: &ArgNames) -> (Vec<Error>, Vec<BadName>) {
+    let mut errors = Vec::new();
+    let mut bad_names = Vec::new();
```

Comment:
Pass in these as mutable references to avoid allocating these vectors and then immediately emptying them into another vector